### PR TITLE
Fix bug with NC tetrahedra in ParNCMesh::GetFaceNeighbors

### PR DESCRIPTION
**Pull request by hughcars**
This PR is an attempt solicit feedback on a fix for https://github.com/mfem/mfem/issues/2666. It includes a test to attempt to verify a serial and parallel solve are identical, and that `ExchangeFaceNbrData()` does not fail `MFEM_ASSERT` or go out of bounds accessing point matrices on faces.

I am particularly looking for feedback on:
1. My conclusion that no additional modification of the point matrix for the degenerate face is required, in the case of a face-edge constraint being shared. This was based on comparing to the `Dim==2` case for an actual slave-master edge in the processor boundary.
2. Whether there is a better test for the impact of the face-edge constraint being communicated incorrectly. 

I used a mass matrix inversion to avoid the need for boundary conditions and for the simplicity. I am not satisfied with this entirely however, as the failure is only triggered by the call to `ExchangeFaceNbrData()`. It does demonstrate that results are the same calculated on the partitioned mesh and the serial mesh though, so it may be sufficient to demonstrate that this "do nothing" solution is right, as there wasn't anything wrong aside from the bad access. 

An alternative test might consider some functional over inter element boundaries, for instance computing inter element jumps for an L2 interpolant. I had not implemented such a test yet as I wasn't aware of a simple way to do so, but figured feedback might point me in the correct direction if that was desired.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3587](https://github.com/mfem/mfem/pull/3587) | @hughcars | @mlstowell | @mlstowell + @jakubcerveny + @dylan-copeland | 5/9/23 | 5/17/23 | ⌛due 5/30/23 | |

### CODE REVIEW

**Comment by hughcars** [comment]
@jakubcerveny Just wondering if you might have input on this as the original architect of the face-edge constraint fix?

File: mesh/pncmesh.cpp [hunk 1]
```diff
@@ -972,8 +985,8 @@ void ParNCMesh::GetFaceNeighbors(ParMesh &pmesh)
    std::map<int, int> vert_map;
    for (int i = 0; i < fnbr.Size(); i++)
    {
-      Element* elem = fnbr[i];
-      mfem::Element* fne = NewMeshElement(elem->geom);
+      auto* elem = fnbr[i];
```
Comment (jakubcerveny) [inline mesh/pncmesh.cpp:988]: I suggest reverting this change is it hurts readability. Especially here, where it's good to know explicitly that `elem` is the internal NCMesh element and `fne` is the regular MFEM element.

File: mesh/pncmesh.cpp [hunk 2]
```diff
@@ -1001,10 +1014,10 @@ void ParNCMesh::GetFaceNeighbors(ParMesh &pmesh)
       if (coordinates.Size())
       {
          tmp_vertex = new TmpVertex[nodes.NumIds()]; // TODO: something cheaper?
-         for (auto it = vert_map.begin(); it != vert_map.end(); ++it)
+         for (const auto &v : vert_map)
```
Comment (jakubcerveny) [inline mesh/pncmesh.cpp:1017]: This is good

File: mesh/pncmesh.cpp [hunk 3]
```diff
@@ -1031,15 +1044,13 @@ void ParNCMesh::GetFaceNeighbors(ParMesh &pmesh)
...
+   for (const auto& cf : shared.conforming)
    {
-      const MeshId &cf = shared.conforming[i];
       Face* face = GetFace(elements[cf.element], cf.local);
-
       Element* e[2] = { &elements[face->elem[0]], &elements[face->elem[1]] };
       if (e[0]->rank == MyRank) { std::swap(e[0], e[1]); }
 
-      Mesh::FaceInfo &fi = pmesh.faces_info[cf.index];
+      auto &fi = pmesh.faces_info[cf.index];
```
Comment (jakubcerveny) [inline mesh/pncmesh.cpp:1053]: But this again makes the reader do more work if they need to figure out the type.

**Review by jakubcerveny — APPROVED** [review]
Thanks a lot for fixing this and for adding the unit test. I'm afraid I can no longer tell authoritatively if this is all that's needed, but having the test is reassuring. I think it's a great idea and I wish we had more of NCMesh covered, like the P matrix construction. I tried implementing and automated test in the past but it wasn't too easy and I gave up. The idea was to evaluate the solution near element boundaries, in the physical domain (thus also testing the reference mappings), left and right off the interface, and then comparing. Doing it on the parallel interfaces would be even more useful, but I couldn't find an elegant way to do it.

File: mesh/pncmesh.cpp [hunk 4]
```diff
@@ -1088,8 +1100,14 @@ void ParNCMesh::GetFaceNeighbors(ParMesh &pmesh)
 
             bool sloc = (sfe.rank == MyRank);
             bool mloc = (mfe.rank == MyRank);
-            if (sloc == mloc) { continue; }
+            if (sloc == mloc // both or neither face is owned by this processor
+                || sf.index < 0) // the face is degenerate (i.e. a face-edge constraint)
+            {
+               continue;
+            }
```
Comment (termi-official) [inline mesh/pncmesh.cpp:1107]: Sorry, I have to ask, but any idea why this should work for hexahedra, but not for tetrahedra? Could you try for higher order space if this still gives the correct results?

File: mesh/pncmesh.cpp [hunk 5]
```diff
@@ -1119,12 +1137,18 @@ void ParNCMesh::GetFaceNeighbors(ParMesh &pmesh)
...
-               std::swap((*pm2)(1,1), (*pm2)(1,3));
+               if (sf.geom == Geometry::Type::SQUARE)
+               {
+                  std::swap((*pm2)(0, 1), (*pm2)(0, 3));
+                  std::swap((*pm2)(1, 1), (*pm2)(1, 3));
+               }
+               else if (sf.geom == Geometry::Type::TRIANGLE)
+               {
+                  std::swap((*pm2)(0, 0), (*pm2)(0, 1));
+                  std::swap((*pm2)(1, 0), (*pm2)(1, 1));
```
Comment (termi-official) [inline mesh/pncmesh.cpp:1150]: I think this flip is incorrect. Can we add unit tests for these? (Test coverage seems to be missing already for the Hex case)

Comment (hughcars) [reply mesh/pncmesh.cpp:1150]: I had been working with the face orientation being 0 as defined on [this line](https://github.com/mfem/mfem/pull/3587/files#diff-c43ff7c21ae1304f127c667baab4061293c38b46496d8603c7e291bc9f6cbac7R1114-R1115). Then going to `geom.cpp`, to swap from orientation 0 to 1, requires swapping the first two columns, assuming the same even-odd patterning as discussed [on the ticket](https://github.com/mfem/mfem/issues/2666#issuecomment-984224259). 

I wasn't sure of how to meaningfully add unit tests past the NCMesh test I put in here, which i acknowledge isn't very satisfactory either. A similar one could be done for hexes, but addressing that wasn't the goal of this PR, as I was just wanting to fix up tets.

File: mesh/pncmesh.cpp (hunk 4 above)
Comment (hughcars) [reply mesh/pncmesh.cpp:1107]: Re higher order: rerunning the original failing case you reported,
```
mpirun -np 6 ./ex15p -m ../data/escher-p2.mesh -r 1 -est 1 -e 0.00001
```
did not fail for me after these fixes. Additionally I had identified that the same bug had been triggered for P1 meshes previously, so appeared unrelated to the order of the mesh, but the topology.

File: mesh/pncmesh.cpp (hunk 3 above)
Comment (hughcars) [reply mesh/pncmesh.cpp]: fixed in https://github.com/mfem/mfem/pull/3587/commits/1e8ede837ac5f02a7c8dde34b2469311c8e13992

File: mesh/pncmesh.cpp (hunk 4 above)
Comment (termi-official) [reply mesh/pncmesh.cpp:1107]: I see. The solution looks still wrong on this mesh in this branch. I think it might be interesting to see if the error can also be triggered with wedges then (because it should, if it is purely related to the faces). I could not find time to check the logic for wedges yet. However, I am still somewhat suspecting that the serial implementation might contain something spurious. I always had trouble with the Kelly error estimator (basically the jump in the fluxes over the all faces of an element) with tets, even in serial, but I could not find time to look into the details there.

File: mesh/pncmesh.cpp (hunk 5 above)
Comment (termi-official) [reply mesh/pncmesh.cpp:1150]: Probably I misunderstood something in the discussion, but I thought that in [this comment](https://github.com/mfem/mfem/issues/2666#issuecomment-984923708) stated that the orientations in MFEM are defined by even/odd pairs of `Constants<T>::Orient`, i.e.
 https://github.com/mfem/mfem/blob/539f663fe6ad3738b6de9615f6c74dfa44927578/fem/geom.cpp#L909-L913
for triangles. My try can be found in this commit: https://github.com/mfem/mfem/commit/ec7a37b167452b272e794580c8b35082d9bde0cd 

Not arguing that you should also add tests for the hex here, sorry for the misunderstanding. I really just want to make sure to get this flip right with some minimal test (to possibly catch regressions), but I also do not have a really good idea how a proper test could look.

**Comment by v-dobrev** [comment]
I'm not 100% sure it will help here but one way to check for consistency across faces is this method: https://github.com/mfem/mfem/blob/1ee0e3b365600fa868dab37a5db5e058d921a929/fem/eltrans.hpp#L598-L616

The idea will be to loop over all faces, get the corresponding `FaceElementTransformations` and check the number returned by the above method.

File: mesh/pncmesh.cpp (hunk 5 above)
Comment (hughcars) [reply mesh/pncmesh.cpp:1150]: I originally had an implementation with the switching, but if you look at the definition of `fi.Elem2Inf` in the lines preceding this section, it is defined as `64 * Elem2No`, i.e. the orientation is always zero after the modulo 64. So I reduced the complexity and just deleted all the switch statements as they couldn't be triggered anyway.

**Comment by hughcars** [comment]
> I'm not 100% sure it will help here but one way to check for consistency across faces is this method:
> 
> https://github.com/mfem/mfem/blob/1ee0e3b365600fa868dab37a5db5e058d921a929/fem/eltrans.hpp#L598-L616
> 
> The idea will be to loop over all faces, get the corresponding `FaceElementTransformations` and check the number returned by the above method.

done in https://github.com/mfem/mfem/pull/3587/commits/627ea87fed782cb24afa7af2bc93fe8ddb273899

**Comment by mlstowell** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: tests/unit/mesh/test_ncmesh.cpp [hunk 6]
```diff
@@ -297,6 +297,228 @@ TEST_CASE("pNCMesh PA diagonal",  "[Parallel], [NCMesh]")
...
+      {
+         const auto face_transform = ttmp.GetFaceElementTransformations(iface);
+
+         CHECK(std::abs(face_transform->CheckConsistency(0)) < 1e-12);
+      }
+
+      // Use P4 to ensure there's a few fully interior DOF.
+      check_l2_projection(ttmp, sttmp, 4);
+
+      ttmp.ExchangeFaceNbrData(); // This causes issues.
```
Comment (dylan-copeland) [inline tests/unit/mesh/test_ncmesh.cpp:514]: Is this comment obsolete? Doesn't this PR resolve the issues?

File: tests/unit/mesh/test_ncmesh.cpp [hunk 7]
```diff
@@ -297,6 +297,228 @@ TEST_CASE("pNCMesh PA diagonal",  "[Parallel], [NCMesh]")
...
+
+   REQUIRE(pmesh.GetGlobalNE() == 8 + 1);
+   REQUIRE(smesh.GetNE() == 8 + 1);
+
+   // This is the set of (i,j) that result in face-edge constraints.
+   // These pairs were arrived at by looping over all possible i,j pairs and
+   // checking for the addition of a face-edge constraint.
+   std::vector<std::pair<int,int>> indices{{2,13}, {3,13}, {6,2}, {6,3}};
+
+   // Rank 0 has all but one element.
```
Comment (dylan-copeland) [inline tests/unit/mesh/test_ncmesh.cpp:457]:
```suggestion
   // Rank 0 has all but one element, in parallel with more than 1 rank.
```

File: tests/unit/mesh/test_ncmesh.cpp [hunk 8]
```diff
@@ -297,6 +297,228 @@ TEST_CASE("pNCMesh PA diagonal",  "[Parallel], [NCMesh]")
...
+         pcg.SetTol(linear_tol);
+         pcg.SetMaxIter(500);
+         pcg.SetPrintLevel(-1);
+         pcg.SetPreconditioner(amg);
+         pcg.Mult(B, X);
+         a.RecoverFEMSolution(X, b, x);
+         return x.ComputeL2Error(rhs_coef);
+      }();
+
+      constexpr double test_tol = 1e-9;
```
Comment (dylan-copeland) [inline tests/unit/mesh/test_ncmesh.cpp:417]: The values of `serror` and `perror` are on the order of 1e-4, so I wonder why their difference is on the order of 1e-10. Maybe it is due to the PCG solver used in my build without suitesparse. Did anyone try with suitesparse and get more digits of accuracy? This doesn't have to be answered for the PR review, just curious.

File: tests/unit/mesh/test_ncmesh.cpp [hunk 9]
```diff
@@ -297,6 +297,228 @@ TEST_CASE("pNCMesh PA diagonal",  "[Parallel], [NCMesh]")
...
+      d[0] = -0.5; d[1] = -1; d[2] = -2; // arbitrary
+      d -= x;
+      return std::sin(d * d);
+   };
+
+   // Given a parallel and a serial mesh, perform an L2 projection and check the
+   // solutions match exactly.
+   auto check_l2_projection = [&exact_soln](ParMesh& pmesh, Mesh& smesh, int order)
+   {
+      MPI_Barrier(MPI_COMM_WORLD); // Get our ducks in a row.
```
Comment (dylan-copeland) [inline tests/unit/mesh/test_ncmesh.cpp:328]: This test has two of these `MPI_Barrier` calls, and I confirmed they are unnecessary, by commenting them out and running the test in parallel. It is good to be cautious, but for high-level MFEM code like this test, the MPI support inside the classes should be so robust that `MPI_Barrier` is not necessary to prevent MPI disasters. Although it doesn't make any practical difference, I would prefer not to include `MPI_Barrier` calls in high-level examples like this, since it may give the impression of being necessary, which incorrectly implies that MFEM users have to be very careful about MPI in simple examples.

File: tests/unit/mesh/test_ncmesh.cpp [hunk 10]
```diff
@@ -297,6 +297,228 @@ TEST_CASE("pNCMesh PA diagonal",  "[Parallel], [NCMesh]")
...
+      Mesh sttmp(stmp);
+      sttmp.GeneralRefinement(serial_refines);
+      REQUIRE(sttmp.GetNE() == 1 + 8 - 1 + 8 - 1 + 8); // 23 elements
+
+      // Loop over interior faces, fill and check face transform on the serial.
+      for (int iface = 0; iface < sttmp.GetNumFaces(); ++iface)
+      {
+         const auto face_transform = sttmp.GetFaceElementTransformations(iface);
+
+         CHECK(std::abs(face_transform->CheckConsistency(0)) < 1e-12);
```
Comment (dylan-copeland) [inline tests/unit/mesh/test_ncmesh.cpp:501]: `CheckConsistency` returns a non-negative number, so the `abs` is unnecessary.

File: tests/unit/mesh/test_ncmesh.cpp (hunk 9 above)
Comment (hughcars) [reply tests/unit/mesh/test_ncmesh.cpp:328]: Sounds good. I had originally sprinkled a few of them around during development as I was working with the `GetGlobalElementNumber` method, which if not careful does cause MPI disasters as it had a global communication. Basically had been calling it after mpi ranks had already been branched by an if. I'll get rid of them now though, as they clearly are now redundant.

File: tests/unit/mesh/test_ncmesh.cpp (hunk 7 above)
Comment (hughcars) [reply tests/unit/mesh/test_ncmesh.cpp]: Went with slightly different wording, but hopefully addressed in https://github.com/mfem/mfem/pull/3587/commits/c790cf942bb329d9b90a6ca1de6b31fccbc1d7d9

File: tests/unit/mesh/test_ncmesh.cpp [hunk 11]
```diff
@@ -297,6 +297,228 @@ TEST_CASE("pNCMesh PA diagonal",  "[Parallel], [NCMesh]")
...
+   MPI_Bcast(&serial_refines[0], 1, MPI_INT, refining_rank, MPI_COMM_WORLD);
+
+   // Rank 0 refines the parallel mesh, all ranks refine the serial mesh
+   smesh.GeneralRefinement(serial_refines, 1); // nonconformal
+   pmesh.GeneralRefinement(refines, 1); // nonconformal
+
+   REQUIRE(pmesh.GetGlobalNE() == 8 + 1);
+   REQUIRE(smesh.GetNE() == 8 + 1);
+
+   // This is the set of (i,j) that result in face-edge constraints.
```
[the remark this answers is not in the document]
Comment (hughcars) [reply tests/unit/mesh/test_ncmesh.cpp]: Addressed in https://github.com/mfem/mfem/pull/3587/commits/c790cf942bb329d9b90a6ca1de6b31fccbc1d7d9

File: tests/unit/mesh/test_ncmesh.cpp (hunk 6 above)
Comment (hughcars) [reply tests/unit/mesh/test_ncmesh.cpp]: Resolved in https://github.com/mfem/mfem/pull/3587/commits/c790cf942bb329d9b90a6ca1de6b31fccbc1d7d9

File: mesh/pncmesh.cpp [hunk 12]
```diff
@@ -893,7 +893,15 @@ void ParNCMesh::GetFaceNeighbors(ParMesh &pmesh)
...
 
-   int bound = shared.conforming.Size() + shared.slaves.Size();
+   // Counts the number of slave faces of a master. This may be larger than the
+   // number of shared slaves if there exist degenerate slave-faces from face-edge constraints.
+   auto count_slaves = [&](int i, const Master& x)
+   {
+      return i + (x.slaves_end - x.slaves_begin);
+   };
+
+   const int bound = shared.conforming.Size() + std::accumulate(
```
Comment (mlstowell) [inline mesh/pncmesh.cpp:903]: AppVeyor is failing on `std::accumulate`. I suspect we need to explicitly include `<numeric>` here.

**Review by dylan-copeland — APPROVED** [review]

File: mesh/pncmesh.cpp (hunk 12 above)
Comment (hughcars) [reply mesh/pncmesh.cpp:904]: Sorry I missed that, I admit with all the unrelated CI failures I had stopped looking. Fixed in https://github.com/mfem/mfem/pull/3587/commits/58db0c0d65ed78228fce4980c07fc084f0deb88c

**Review by mlstowell — APPROVED** [review]

File: mesh/pncmesh.cpp (hunk 5 above)
Comment (termi-official) [reply mesh/pncmesh.cpp:1151]: I thought it is set this way because the swap in these lines guarantees that we have this orientation (see the code comment below in this file)?

Btw can we for future purposes add an else with an error message here?
